File: app/MicroMouseDashboard/Task.swift
enum Task: UInt8, CaseIterable, Identifiable {
  case none = 0
  
  // 1-10: Maze tasks
  
  case mazeSearch = 1
  case mazeSlowSolve = 2
  case mazeFastSolve = 3

  // 11-40: Test drive tasks.
  
  case testDriveStraightFromBackWallToSenseSpot = 11
  case testDriveStraightOneCell = 12
  case testDriveTurnRightFromSenseSpotToSenseSpot = 13
  case testDriveTurnLeftFromSenseSpotToSenseSpot = 14
  case testDriveTurnRightInPlace = 15
  case testDriveTurnLeftInPlace = 16
  case testDriveTurn180InPlace = 17
  case testGyro = 18
  case testDriveStraightFourCellsFromBackWallWithVisionAlign = 19
  
  // 41-50: Manual control tasks.
  
  case manualChassisSpeeds = 41

  // 100+: Other

  case armed = 100
  case armedTriggering
  case armedTriggered

  case visionCalibrate
  
  case unknown = 255

  var id: Self { self }
}

let TaskNames: [Task: String] = [
  .none: "None",
  
  .mazeSearch: "Maze Search",
  .mazeSlowSolve: "Maze Slow Solve",
  .mazeFastSolve: "Maze Fast Solve",

  .testDriveStraightFromBackWallToSenseSpot: "TEST - Drive Straight From Back Wall To Sense Spot",
  .testDriveStraightOneCell: "TEST - Drive Straight One Cell",
  .testDriveTurnRightFromSenseSpotToSenseSpot: "TEST - Drive Turn Right From Sense Spot To Sense Spot",
  .testDriveTurnLeftFromSenseSpotToSenseSpot: "TEST - Drive Turn Left From Sense Spot To Sense Spot",
  .testDriveTurnRightInPlace: "TEST - Drive Turn Right In Place",
  .testDriveTurnLeftInPlace: "TEST - Drive Turn Left In Place",
  .testDriveTurn180InPlace: "TEST - Drive Turn 180 In Place",
  .testGyro: "TEST - Gyro",
  .testDriveStraightFourCellsFromBackWallWithVisionAlign: "TEST - Drive Straight Four Cells From Back Wall With Vision Align",
  
  .manualChassisSpeeds: "Manual Control - Chassis Speeds",

  .armed: "Armed",
  .armedTriggering: "Armed Triggering",
  .armedTriggered: "Armed Triggered",

  .visionCalibrate: "Vision Calibrating",
  
  .unknown: "Unknown",
]

let TaskDescriptions: [Task: String] = [
  .none: "None",
  
  .mazeSearch: "Search to the center of the maze, then back to the start",
  .mazeSlowSolve: "Solve the maze using the same control method as Search mode",
  .mazeFastSolve: "Solve the maze as fast as possible while using previous search data",

  .testDriveStraightFromBackWallToSenseSpot: "Drive straight from back wall to sense spot",
  .testDriveStraightOneCell: "Drive straight for one cell",
  .testDriveTurnRightFromSenseSpotToSenseSpot: "Make a right turn from sense spot to sense spot",
  .testDriveTurnLeftFromSenseSpotToSenseSpot: "Make a left turn from sense spot to sense spot",
  .testDriveTurnRightInPlace: "Turn right in place",
  .testDriveTurnLeftInPlace: "Turn left in place",
  .testDriveTurn180InPlace: "Turn 180 degrees in place",
  .testGyro: "Maintain a rotational velocity of 0 deg/s",
  .testDriveStraightFourCellsFromBackWallWithVisionAlign: "Drive straight four cells from back wall with vision align",
]

File: firmware/lib/robot/task.cpp
#include <micromouse/robot/task.hpp>

namespace robot {

const char* task_to_string(Task task) {
  switch (task) {
    using enum Task;
    case STOPPED:
      return "Stopped";
    case MAZE_SEARCH:
      return "[Maze] Search";
    case MAZE_SLOW_SOLVE:
      return "[Maze] Slow Solve";
    case MAZE_FAST_SOLVE:
      return "[Maze] Fast Solve";
    case MAZE_SEARCH_START_STOP_MOTION:
      return "[Maze] Search (Start+Stop Motion)";
    case MAZE_SEARCH_TWO_TIMES:
      return "[Maze] Search (Two Times)";
    case MAZE_SEARCH_TWO_TIMES_START_STOP_MOTION:
      return "[Maze] Search (Two Times) (Start+Stop Motion)";
    case MAZE_SOLVE_WITH_SEARCH_NAVIGATION:
      return "[Maze] Solve with Search Navigation";
    case MAZE_SOLVE_WITH_SEARCH_NAVIGATION_START_STOP_MOTION:
      return "[Maze] Solve with Search Navigation (Start+Stop Motion)";
    case MAZE_SEARCH_FASTER:
      return "[Maze] Search (Faster)";
    case MAZE_SEARCH_THEN_SLOW_SOLVE_THEN_FAST_SOLVE:
      return "[Maze] Search then slow solve then fast solve";
    case TEST_DRIVE_STRAIGHT_FROM_BACK_WALL_TO_SENSE_SPOT:
      return "[Test] Drive Straight from Back Wall to Sense Spot";
    case TEST_DRIVE_STRAIGHT_ONE_CELL:
      return "[Test] Drive Straight One Cell";
    case TEST_DRIVE_TURN_RIGHT_FROM_SENSE_SPOT_TO_SENSE_SPOT:
      return "[Test] Drive Turn Right from Sense Spot to Sense Spot";
    case TEST_DRIVE_TURN_LEFT_FROM_SENSE_SPOT_TO_SENSE_SPOT:
      return "[Test] Drive Turn Left from Sense Spot to Sense Spot";
    case TEST_DRIVE_TURN_RIGHT_IN_PLACE:
      return "[Test] Drive Turn Right In Place";
    case TEST_DRIVE_TURN_LEFT_IN_PLACE:
      return "[Test] Drive Turn Left In Place";
    case TEST_DRIVE_TURN_180_IN_PLACE:
      return "[Test] Drive Turn 180 In Place";
    case DRIVE_BACKUP_INTO_WALL:
      return "[Drive] Backup into wall";
    case TEST_GYRO:
      return "[Test] Gyro";
    case TEST_DRIVE_STRAIGHT_FOUR_CELLS_FROM_BACK_WALL_WITH_VISION_ALIGN:
      return "[Test] Drive Straight Four Cells from Back Wall with Vision Align";
    case TEST_DRIVE_RAW_SPEEDS:
      return "[Test] Drive Raw Speeds";
    case TEST_DRIVE_CONSTANT_SPEED:
      return "[Test] Drive Constant Speed";
    case MANUAL_CHASSIS_SPEEDS:
      return "Manual Chassis Speeds";
    case IDLE:
      return "Idle";
    case ARMED:
      return "[Armed]";
    case ARMED_TRIGGERING:
      return "[Armed] Triggering";
    case ARMED_TRIGGERED:
      return "[Armed] Triggered";
    case VISION_CALIBRATE:
      return "[Vision] Calibrate";
  }
  return "Unknown";
}

}  // namespace robot
